<a href="https://colab.research.google.com/github/LaviniBastos/Analise-Exploratoria-de-Dados/blob/main/Dados_Faltantes_Dados_Constantes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lidar com dados no mundo real, não é uma tarefa muito fácil, precisamos entender e saber lidar com dados ruins. Os dois tipos de dados ruins mais comuns em uma análise de dados são os dados **constantes** e os dados **errôneos**

O **DataSet** Utilizado aqui será referente a casas em São Paulo.

In [1]:
import numpy as np
import pandas as pd

In [46]:
casas_sp = pd.read_csv('/houses_sp.csv')

In [47]:
casas_sp.head(10)

city   area  rooms  bathroom  parking spaces floor   hoa   rent   tax  \
0  São Paulo   70.0    2.0         1             1.0     7  2065   3300   211   
1  São Paulo  320.0    4.0         4             2.0    20  1200   4960  1750   
2  São Paulo   25.0    1.0         1             NaN     1     0    800    25   
3  São Paulo  650.0    3.0         3             7.0     -     0   8000   834   
4  São Paulo  213.0    NaN         4             4.0     4  2254   3223  1735   
5  São Paulo  152.0    2.0         2             1.0     3  1000  15000   250   
6  São Paulo   26.0    1.0         1             NaN     2   470   2100   150   
7  São Paulo   36.0    1.0         1             NaN    11   359   2100    70   
8  São Paulo   55.0    1.0         1             1.0     2   790   4200   224   
9  São Paulo  100.0    2.0         2             2.0    24   900   4370    17   

   insurance    total  page hits  days available  interactions  \
0         42   5618.0        324              23           108   
1         63   7973.0        720              78           240   
2         11    836.0       1548              78           516   
3        121   8955.0        396              66           132   
4         41   7253.0        756              99           252   
5        191  16441.0       2412             142           804   
6         27   2747.0        828              85           276   
7         27   2556.0        360              22           120   
8         54   5268.0        540             120           180   
9         56   5343.0       1044              65           348   

   weeks available   type  
0                3   flat  
1               11   flat  
2               11   flat  
3                9  house  
4               14   flat  
5               20   flat  
6               12   flat  
7                3   flat  
8               17   flat  
9                9   flat

Pelo output podemos ver que as informações de casas disponíveis pertencem somente a cidade de São Paulo, então a coluna 'city' é irrelevante para nós neste contexto, mas vamos avaliar com mais profundidade


In [49]:
casas_sp[('city')].value_counts()

city
São Paulo    7143
Name: count, dtype: int64

In [50]:
casas_sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7143 entries, 0 to 7142
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   city             7143 non-null   object 
 1   area             7108 non-null   float64
 2   rooms            7092 non-null   float64
 3   bathroom         7143 non-null   int64  
 4   parking spaces   5586 non-null   float64
 5   floor            7143 non-null   object 
 6   hoa              7143 non-null   int64  
 7   rent             7143 non-null   int64  
 8   tax              7143 non-null   object 
 9   insurance        7143 non-null   int64  
 10  total            7143 non-null   float64
 11  page hits        7143 non-null   int64  
 12  days available   7143 non-null   int64  
 13  interactions     7143 non-null   int64  
 14  weeks available  7143 non-null   int64  
 15  type             7143 non-null   object 
dtypes: float64(4), int64(8), object(4)
memory usage: 893.0+ KB


O **DataSet** possui *7143 entradas* e todas estão no contexto de São Paulo, então podemos remover essa coluna, já que ela não vai interferir nos resultados. Esse é considerado um tipo de dado **"constante"**, e ele é um problema pois pode ocasionar um custo computacional grande nas análises, principalmente se o DataSet contiver alguns milhões de linhas.

In [51]:
casas_sp = casas_sp.drop(columns=['city'])

In [52]:
casas_sp.head()

area  rooms  bathroom  parking spaces floor   hoa  rent   tax  insurance  \
0   70.0    2.0         1             1.0     7  2065  3300   211         42   
1  320.0    4.0         4             2.0    20  1200  4960  1750         63   
2   25.0    1.0         1             NaN     1     0   800    25         11   
3  650.0    3.0         3             7.0     -     0  8000   834        121   
4  213.0    NaN         4             4.0     4  2254  3223  1735         41   

    total  page hits  days available  interactions  weeks available   type  
0  5618.0        324              23           108                3   flat  
1  7973.0        720              78           240               11   flat  
2   836.0       1548              78           516               11   flat  
3  8955.0        396              66           132                9  house  
4  7253.0        756              99           252               14   flat

Olhando novamente para o **DataSet** podemos perceber também que existe registros denominado "NaN" *(Nulos).* Esses são os dados *faltantes*. E isso é um problema pois se precisarmos calcular uma média, esse dado será desconsiderado e pode influênciar na análise.

In [53]:
casas_sp['rooms']

0       2.0
1       4.0
2       1.0
3       3.0
4       NaN
       ... 
7138    2.0
7139    4.0
7140    3.0
7141    3.0
7142    4.0
Name: rooms, Length: 7143, dtype: float64

In [54]:
casas_sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7143 entries, 0 to 7142
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   area             7108 non-null   float64
 1   rooms            7092 non-null   float64
 2   bathroom         7143 non-null   int64  
 3   parking spaces   5586 non-null   float64
 4   floor            7143 non-null   object 
 5   hoa              7143 non-null   int64  
 6   rent             7143 non-null   int64  
 7   tax              7143 non-null   object 
 8   insurance        7143 non-null   int64  
 9   total            7143 non-null   float64
 10  page hits        7143 non-null   int64  
 11  days available   7143 non-null   int64  
 12  interactions     7143 non-null   int64  
 13  weeks available  7143 non-null   int64  
 14  type             7143 non-null   object 
dtypes: float64(4), int64(8), object(3)
memory usage: 837.2+ KB


Pelo output é possível ver que, na coluna rooms "7092" Registros estão preenchidos.
O restante para chegar na quantidade total estão nulos.

Para tratar esses dados nulos de uma forma onde não perderiamos os outros registros associados a linha dele *(se optassemos por simplesmente excluir)*, seria calcular a mediana da coluna *"rooms"* e preencher  os dados nulos com o resultado dessa mediana

*A média nesses casos não são recomendadas pois elas podem sofrer discrepâncias se houvermos dados muito fora do padrão geral (os outliers)*

In [55]:
casas_sp['rooms'].median()

3.0

No Geral temos 3 quartos

Agora devemos preencher os dados nulos com uma função do pandas chamada: **fillna()**

A própria função calculou a mediana e preencheu os dados nulos com ela.
Mas precisamos substituir a coluna antiga pela nova tratada, é importante definir a coluna **'rooms'** na variável nova, se não todos os dados serão substituidos.

In [56]:
casas_sp['rooms'] = casas_sp['rooms'].fillna(casas_sp['rooms'].median())

In [57]:
casas_sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7143 entries, 0 to 7142
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   area             7108 non-null   float64
 1   rooms            7143 non-null   float64
 2   bathroom         7143 non-null   int64  
 3   parking spaces   5586 non-null   float64
 4   floor            7143 non-null   object 
 5   hoa              7143 non-null   int64  
 6   rent             7143 non-null   int64  
 7   tax              7143 non-null   object 
 8   insurance        7143 non-null   int64  
 9   total            7143 non-null   float64
 10  page hits        7143 non-null   int64  
 11  days available   7143 non-null   int64  
 12  interactions     7143 non-null   int64  
 13  weeks available  7143 non-null   int64  
 14  type             7143 non-null   object 
dtypes: float64(4), int64(8), object(3)
memory usage: 837.2+ KB


Podemos ver que agora a **coluna room** possui todas as 7143 linhas preenchidas

Outra coisa que pode vir a acontecer, é o **DataSet** possuir dados que não fazem muito sentido, são os dados **"Errôneos"**

In [59]:
casas_sp.head(10)

area  rooms  bathroom  parking spaces floor   hoa   rent   tax  insurance  \
0   70.0    2.0         1             1.0     7  2065   3300   211         42   
1  320.0    4.0         4             2.0    20  1200   4960  1750         63   
2   25.0    1.0         1             NaN     1     0    800    25         11   
3  650.0    3.0         3             7.0     -     0   8000   834        121   
4  213.0    3.0         4             4.0     4  2254   3223  1735         41   
5  152.0    2.0         2             1.0     3  1000  15000   250        191   
6   26.0    1.0         1             NaN     2   470   2100   150         27   
7   36.0    1.0         1             NaN    11   359   2100    70         27   
8   55.0    1.0         1             1.0     2   790   4200   224         54   
9  100.0    2.0         2             2.0    24   900   4370    17         56   

     total  page hits  days available  interactions  weeks available   type  
0   5618.0        324              23           108                3   flat  
1   7973.0        720              78           240               11   flat  
2    836.0       1548              78           516               11   flat  
3   8955.0        396              66           132                9  house  
4   7253.0        756              99           252               14   flat  
5  16441.0       2412             142           804               20   flat  
6   2747.0        828              85           276               12   flat  
7   2556.0        360              22           120                3   flat  
8   5268.0        540             120           180               17   flat  
9   5343.0       1044              65           348                9   flat

A coluna floor, tem um registro "-", o que não nos diz nada, mas no contexto dos dados trata-se de uma casa, que não está em nenhum andar, deveria estar 0 no lugar do traço. Mas primeiro vamos olhar a frequência desse dado.

In [60]:
casas_sp['floor'].value_counts()

floor
-      1847
1       727
2       506
4       402
3       399
5       370
7       312
6       297
8       291
10      256
11      241
9       230
14      206
13      206
12      190
15      120
16       84
17       71
21       58
20       56
18       56
26       54
19       45
25       36
22       22
23       21
24       18
27        7
28        6
29        4
32        2
35        1
110       1
46        1
Name: count, dtype: int64

A frequência dele é bem grande e não tem como ser ignorada, precisamos substituir pois se precisarmos fazer calculos *com a média dos andares mais populares por exemplo* isso irá influênciar no resultado.

In [61]:
casas_sp.loc[casas_sp['floor'] == '-', 'floor']

3       -
11      -
12      -
14      -
23      -
       ..
7114    -
7123    -
7125    -
7133    -
7134    -
Name: floor, Length: 1847, dtype: object

In [62]:
casas_sp.loc[casas_sp['floor'] == '-', 'floor'] = '0'

In [63]:
casas_sp.loc[casas_sp['floor'] == '0', 'floor']

3       0
11      0
12      0
14      0
23      0
       ..
7114    0
7123    0
7125    0
7133    0
7134    0
Name: floor, Length: 1847, dtype: object

Os **tracinhos** foram substituidos por **zeros**, que representam melhor uma casa que se encontra no térreo.